File: rocketmq-go/model/message/message_queue.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package message

type MessageQueue struct {
	topic      string
	brokerName string
	queueId    int32
}

func NewMessageQueue(topic string, brokerName string, queueId int32) *MessageQueue {
	return &MessageQueue{
		topic:      topic,
		brokerName: brokerName,
		queueId:    queueId,
	}
}

func (queue *MessageQueue) clone() *MessageQueue {
	no := new(MessageQueue)
	no.topic = queue.topic
	no.queueId = queue.queueId
	no.brokerName = queue.brokerName
	return no
}

func (queue MessageQueue) BrokerName() string {
	return queue.brokerName
}

func (queue *MessageQueue) QueueID() int32 {
	return queue.queueId
}

type MessageQueues []*MessageQueue

func (queues MessageQueues) Less(i, j int) bool {
	imq := queues[i]
	jmq := queues[j]

	if imq.topic < jmq.topic {
		return true
	}

	if imq.topic < jmq.topic {
		return false
	}

	if imq.brokerName < jmq.brokerName {
		return true
	}

	if imq.brokerName < jmq.brokerName {
		return false
	}

	if imq.queueId < jmq.queueId {
		return true
	}

	return false
}

func (queues MessageQueues) Swap(i, j int) {
	queues[i], queues[j] = queues[j], queues[i]
}

func (queues MessageQueues) Len() int {
	return len(queues)
}

File: rocketmq-client-php/src/message_queue.cc
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

#include "message_queue.h"

MessageQueue::MessageQueue(const rocketmq::MQMessageQueue& other){
    messageQueue = other;
}

void MessageQueue::__construct(Php::Parameters &params){
    std::string topic = params[0];
    std::string brokerName = params[1];
    int queueId = params[2];

    this->messageQueue = rocketmq::MQMessageQueue(topic, brokerName, queueId);
}

Php::Value MessageQueue::getTopic(){
    return this->messageQueue.getTopic();
}

void MessageQueue::setTopic(Php::Parameters &param){
    std::string topic = param[0];
    this->messageQueue.setTopic(topic);
}

Php::Value MessageQueue::getBrokerName(){
    return this->messageQueue.getBrokerName();
}

void MessageQueue::setBrokerName(Php::Parameters &param){
    std::string brokerName = param[0];
    this->messageQueue.setBrokerName(brokerName);
}

Php::Value MessageQueue::getQueueId() {
    return this->messageQueue.getQueueId();
}

void MessageQueue::setQueueId(Php::Parameters &param){
    return this->messageQueue.setQueueId(param[0]);
}

void registerMessageQueue(Php::Namespace &rocketMQNamespace){
    Php::Class<MessageQueue> messageQueueClass("MessageQueue");

    messageQueueClass.method<&MessageQueue::getTopic>("getTopic");
    messageQueueClass.method<&MessageQueue::setTopic>("setTopic", { Php::ByVal("topic", Php::Type::String), });

    messageQueueClass.method<&MessageQueue::getBrokerName>("getBrokerName");
    messageQueueClass.method<&MessageQueue::setBrokerName>("setBrokerName", { Php::ByVal("brokerName", Php::Type::String), });

    messageQueueClass.method<&MessageQueue::getQueueId>("getQueueId");
    messageQueueClass.method<&MessageQueue::setQueueId>("setQueueId", { Php::ByVal("queueId", Php::Type::String), });

    messageQueueClass.method<&MessageQueue::__construct>("__construct", {
            Php::ByVal("topic", Php::Type::String),
            Php::ByVal("brokerName", Php::Type::String),
            Php::ByVal("queueId", Php::Type::Numeric),
            });

    rocketMQNamespace.add(messageQueueClass);
}
